<a href="https://colab.research.google.com/github/kochlisGit/Advanced-ML/blob/main/Categorical_Encoding.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [4]:
!pip install tensorflow_addons
!pip install category_encoders

In [5]:
import warnings
warnings.filterwarnings("ignore")

In [6]:
import pandas as pd
import requests
from zipfile import ZipFile

dataset_url = 'https://storage.googleapis.com/kaggle-data-sets/225/498/bundle/archive.zip?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com%40kaggle-161607.iam.gserviceaccount.com%2F20220430%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20220430T104007Z&X-Goog-Expires=259199&X-Goog-SignedHeaders=host&X-Goog-Signature=1ced48204b7889c650880becabe6f5a825ff8d6d346832d811cb4a928586f097ff2edd3f976109f5c05b3bcb7c93c644196bbf71417ed9038f23296c2310fda6539c5349471f856435c1ee13f0345cefd37aca6de9f39f454486106353681c949830c5629f62ed7551beb1e16dda1f011b4c54f9c1943e2607629e5b6849373b923fd595fdcb63a6e7a61a0d98c3753ffdafaeb4506efafe45948cb2dc577c2df8d0cf6d195c88077f050e024ffb50f3f66b0f1fa4b0c1fe3ac7c5185aa0af2907d179847e8eec2d4996428fa2b97b93c9d19247827213c65fb142e4d5f3ce20f3ba0b7fa3a45a55a17f8b975f2204e77fcc9edaff701c1c7d29e459ce6f4e25'
download_filename = 'download.zip'
content_filename = 'adult.csv'

req = requests.get(dataset_url)
with open(download_filename, 'wb') as output_file:
  output_file.write(req.content)
print('Download completed!\n')

zf = ZipFile(download_filename)
data = pd.read_csv(zf.open(content_filename)).dropna()
data['occupation'] = data['occupation'].replace({'?': 'Unknown'})
data['workclass'] = data['workclass'].replace({'?': 'Unknown'})
data

Download completed!



age workclass  fnlwgt     education  education.num      marital.status  \
0       90   Unknown   77053       HS-grad              9             Widowed   
1       82   Private  132870       HS-grad              9             Widowed   
2       66   Unknown  186061  Some-college             10             Widowed   
3       54   Private  140359       7th-8th              4            Divorced   
4       41   Private  264663  Some-college             10           Separated   
...    ...       ...     ...           ...            ...                 ...   
32556   22   Private  310152  Some-college             10       Never-married   
32557   27   Private  257302    Assoc-acdm             12  Married-civ-spouse   
32558   40   Private  154374       HS-grad              9  Married-civ-spouse   
32559   58   Private  151910       HS-grad              9             Widowed   
32560   22   Private  201490       HS-grad              9       Never-married   

              occupation   relationship   race     sex  capital.gain  \
0                Unknown  Not-in-family  White  Female             0   
1        Exec-managerial  Not-in-family  White  Female             0   
2                Unknown      Unmarried  Black  Female             0   
3      Machine-op-inspct      Unmarried  White  Female             0   
4         Prof-specialty      Own-child  White  Female             0   
...                  ...            ...    ...     ...           ...   
32556    Protective-serv  Not-in-family  White    Male             0   
32557       Tech-support           Wife  White  Female             0   
32558  Machine-op-inspct        Husband  White    Male             0   
32559       Adm-clerical      Unmarried  White  Female             0   
32560       Adm-clerical      Own-child  White    Male             0   

       capital.loss  hours.per.week native.country income  
0              4356              40  United-States  <=50K  
1              4356              18  United-States  <=50K  
2              4356              40  United-States  <=50K  
3              3900              40  United-States  <=50K  
4              3900              40  United-States  <=50K  
...             ...             ...            ...    ...  
32556             0              40  United-States  <=50K  
32557             0              38  United-States  <=50K  
32558             0              40  United-States   >50K  
32559             0              40  United-States  <=50K  
32560             0              20  United-States  <=50K  

[32561 rows x 15 columns]

To better understand the categorical encoding techniques, let's first print the unique values for each column.

In [7]:
categorical_columns = ['workclass', 'education', 'marital.status', 'occupation', 'relationship', 'race', 'sex', 'native.country']

for column in categorical_columns:
  print('column: {},\tunique values: {}'.format(column, data[column].unique().shape[0]))

column: workclass,	unique values: 9
column: education,	unique values: 16
column: marital.status,	unique values: 7
column: occupation,	unique values: 15
column: relationship,	unique values: 6
column: race,	unique values: 5
column: sex,	unique values: 2
column: native.country,	unique values: 42


Categorical Column  | Unique Values
-------------------|------------------
workclass       | 9
education       | 16
marital.status       |7 
occupation       | 15
relationship       | 6 
race       | 5
sex       | 2
native.country       | 42
income      | 2


# Binary Encoding of Targets

Since there are 2 categories for the **income**: *{<=50K, >50K}*, we can use binary encoding.

In [8]:
targets = data['income'].replace({'<=50K': 0, '>50K': 1}).to_numpy()
data = data.drop(columns=['income'])
targets

array([0, 0, 0, ..., 1, 0, 0])

Splitting dataset to train-test samples.

In [9]:
from sklearn.model_selection import train_test_split

test_size = 0.1
random_state=0

x_train, x_test, y_train, y_test = train_test_split(
    data,
    targets,
    test_size=test_size,
    random_state=random_state
)
x_train.shape, x_test.shape, y_train.shape, y_test.shape

((29304, 14), (3257, 14), (29304,), (3257,))

#Ordinal Encoding

In ordinal encoding, we replace each categorical variable with an integer value (e.g. 0, 1, 2, etc.). In this case, retaining the order is important.

In [11]:
from sklearn.preprocessing import OrdinalEncoder

ordinal_encoder = OrdinalEncoder()
ordinal_encoder.fit(data)
x_train = ordinal_encoder.transform(x_train)
x_test = ordinal_encoder.transform(x_test)
x_train

array([[1.8000e+01, 3.0000e+00, 1.5732e+04, ..., 0.0000e+00, 3.9000e+01,
        3.9000e+01],
       [1.1000e+01, 3.0000e+00, 3.2850e+03, ..., 0.0000e+00, 2.2000e+01,
        3.9000e+01],
       [3.8000e+01, 3.0000e+00, 2.3630e+03, ..., 0.0000e+00, 3.9000e+01,
        3.9000e+01],
       ...,
       [9.0000e+00, 3.0000e+00, 1.6120e+03, ..., 0.0000e+00, 5.0000e+01,
        3.9000e+01],
       [2.7000e+01, 3.0000e+00, 5.1460e+03, ..., 0.0000e+00, 3.9000e+01,
        3.9000e+01],
       [2.2000e+01, 3.0000e+00, 1.4687e+04, ..., 0.0000e+00, 3.9000e+01,
        3.9000e+01]])

Building an MLP (Multi-Layer Perceptron) classifier.

In [10]:
import tensorflow as tf
import tensorflow_addons as tfa


def build_model(input_size):
  model = tf.keras.models.Sequential()
  model.add(tf.keras.layers.Input(input_size))
  model.add(tf.keras.layers.GaussianNoise(stddev=0.1))
  model.add(tf.keras.layers.Dense(units=128, activation='gelu', use_bias=False))
  model.add(tf.keras.layers.BatchNormalization())
  model.add(tf.keras.layers.Dropout(rate=0.2))
  model.add(tf.keras.layers.Dense(units=64, kernel_regularizer='l1', use_bias=False))
  model.add(tf.keras.layers.BatchNormalization())
  model.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

  model.compile(
      optimizer=tfa.optimizers.Yogi(learning_rate=0.001),
      loss=tf.keras.losses.BinaryCrossentropy(label_smoothing=0.1),
      metrics=['accuracy']
  )
  return model


model = build_model(data.shape[1:])
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 gaussian_noise (GaussianNoi  (None, 14)               0         
 se)                                                             
                                                                 
 dense (Dense)               (None, 128)               1792      
                                                                 
 batch_normalization (BatchN  (None, 128)              512       
 ormalization)                                                   
                                                                 
 dropout (Dropout)           (None, 128)               0         
                                                                 
 dense_1 (Dense)             (None, 64)                8192      
                                                                 
 batch_normalization_1 (Batc  (None, 64)               2

Training the neural network with the ordinal variables.

In [ ]:
batch_size = 32
epochs = 20
shuffle=True

model.fit(x=x_train, y=y_train, batch_size=batch_size, epochs=epochs, shuffle=shuffle, validation_data=(x_test, y_test))

Epoch 1/20
916/916 [==============================] - 22s 13ms/step - loss: 1.6441 - accuracy: 0.7641 - val_loss: 0.6491 - val_accuracy: 0.7710
Epoch 2/20
916/916 [==============================] - 7s 8ms/step - loss: 0.6042 - accuracy: 0.7877 - val_loss: 0.6521 - val_accuracy: 0.7731
Epoch 3/20
916/916 [==============================] - 7s 8ms/step - loss: 0.5706 - accuracy: 0.7942 - val_loss: 0.9231 - val_accuracy: 0.5388
Epoch 4/20
916/916 [==============================] - 7s 8ms/step - loss: 0.5486 - accuracy: 0.8001 - val_loss: 0.6819 - val_accuracy: 0.6460
Epoch 5/20
916/916 [==============================] - 7s 7ms/step - loss: 0.5369 - accuracy: 0.8031 - val_loss: 0.5629 - val_accuracy: 0.7928
Epoch 6/20
916/916 [==============================] - 7s 8ms/step - loss: 0.5313 - accuracy: 0.8046 - val_loss: 0.5119 - val_accuracy: 0.8158
Epoch 7/20
916/916 [==============================] - 7s 8ms/step - loss: 0.5266 - accuracy: 0.8032 - val_loss: 0.6646 - val_accuracy: 0.6736
Epoc

Best Validation Accuracy: **0.8330**

#One-Hot Encoding
This cons of the ordinal encoding is that categorical variables become an **ordered** set of numbers. However, this could potentially trick the classifier into thinking that close numbers are similar (e.g. `Married and Divorced` values have the opposite meaning. However, `if Married = 0 and Divorced = 1` is assigned, then it might confuse the classifier into thinking that they are similar.

One-Hot encoding takes care of the above disadvantage by separating the categorical columns into multiple binary columns (e.g. `Sex: {Male, Female}` becomes `s1, s2, where (s1, s2) = (1, 0) if Male and (0, 1) if Female`). This might work for a few unique variables. **However, if the number of unique values is large, then the dimensions of the dataset would become a very large sparsed matrix.**

#Dummy Encoding
This is an improvement over One-Hot Encoding. Dummy encoding uses less N-1 features to represent N categories. For example, `Marital Status: {Married, Divorced, Engaged} becomes s1, s2, where (s1, s2) = (0, 0) if Married, (0, 1) if Divorced and (1, 1) if engaged).` However, in this dataset 

In [ ]:
dummy_data = pd.get_dummies(data)
dummy_data

age  fnlwgt  education.num  capital.gain  capital.loss  hours.per.week  \
0       90   77053              9             0          4356              40   
1       82  132870              9             0          4356              18   
2       66  186061             10             0          4356              40   
3       54  140359              4             0          3900              40   
4       41  264663             10             0          3900              40   
...    ...     ...            ...           ...           ...             ...   
32556   22  310152             10             0             0              40   
32557   27  257302             12             0             0              38   
32558   40  154374              9             0             0              40   
32559   58  151910              9             0             0              40   
32560   22  201490              9             0             0              20   

       workclass_Federal-gov  workclass_Local-gov  workclass_Never-worked  \
0                          0                    0                       0   
1                          0                    0                       0   
2                          0                    0                       0   
3                          0                    0                       0   
4                          0                    0                       0   
...                      ...                  ...                     ...   
32556                      0                    0                       0   
32557                      0                    0                       0   
32558                      0                    0                       0   
32559                      0                    0                       0   
32560                      0                    0                       0   

       workclass_Private  ...  native.country_Portugal  \
0                      0  ...                        0   
1                      1  ...                        0   
2                      0  ...                        0   
3                      1  ...                        0   
4                      1  ...                        0   
...                  ...  ...                      ...   
32556                  1  ...                        0   
32557                  1  ...                        0   
32558                  1  ...                        0   
32559                  1  ...                        0   
32560                  1  ...                        0   

       native.country_Puerto-Rico  native.country_Scotland  \
0                               0                        0   
1                               0                        0   
2                               0                        0   
3                               0                        0   
4                               0                        0   
...                           ...                      ...   
32556                           0                        0   
32557                           0                        0   
32558                           0                        0   
32559                           0                        0   
32560                           0                        0   

       native.country_South  native.country_Taiwan  native.country_Thailand  \
0                         0                      0                        0   
1                         0                      0                        0   
2                         0                      0                        0   
3                         0                      0                        0   
4                         0                      0                        0   
...                     ...                    ...                      ...   
32556                     0                      0                        0   
32557                     0        

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(
    dummy_data,
    targets,
    test_size=test_size,
    random_state=random_state
)

model = build_model(dummy_data.shape[1:])
model.fit(x=x_train, y=y_train, batch_size=batch_size, epochs=epochs, shuffle=shuffle, validation_data=(x_test, y_test))

Epoch 1/20
916/916 [==============================] - 9s 8ms/step - loss: 1.7158 - accuracy: 0.7689 - val_loss: 0.6750 - val_accuracy: 0.7749
Epoch 2/20
916/916 [==============================] - 7s 7ms/step - loss: 0.6385 - accuracy: 0.7785 - val_loss: 0.6733 - val_accuracy: 0.7181
Epoch 3/20
916/916 [==============================] - 7s 8ms/step - loss: 0.6073 - accuracy: 0.7844 - val_loss: 0.6152 - val_accuracy: 0.7706
Epoch 4/20
916/916 [==============================] - 7s 7ms/step - loss: 0.5902 - accuracy: 0.7875 - val_loss: 0.5793 - val_accuracy: 0.7909
Epoch 5/20
916/916 [==============================] - 7s 7ms/step - loss: 0.5867 - accuracy: 0.7870 - val_loss: 0.6048 - val_accuracy: 0.7771
Epoch 6/20
916/916 [==============================] - 7s 7ms/step - loss: 0.5743 - accuracy: 0.7913 - val_loss: 0.5732 - val_accuracy: 0.8044
Epoch 7/20
916/916 [==============================] - 11s 12ms/step - loss: 0.5724 - accuracy: 0.7892 - val_loss: 0.5662 - val_accuracy: 0.7872
Epoc

Best Validation Accuracy: **0.8044**

#Effect/Deviation/Sum Encoding

In dummy coding, we use 0 and 1 to represent the data but in effect encoding (also known as *Deviation Encoding or Sum Encoding*), we use three values i.e. 1,0, and -1.

In [ ]:
from category_encoders.sum_coding import SumEncoder

x_train, x_test, y_train, y_test = train_test_split(
    data,
    targets,
    test_size=test_size,
    random_state=random_state
)

sum_encoder = SumEncoder()
sum_encoder.fit(data)
x_train = sum_encoder.transform(x_train)
x_test = sum_encoder.transform(x_test)
x_train

intercept  age  workclass_0  workclass_1  workclass_2  workclass_3  \
9281           1   35          0.0          1.0          0.0          0.0   
31884          1   28          0.0          1.0          0.0          0.0   
31580          1   55          0.0          1.0          0.0          0.0   
18489          1   33          0.0          1.0          0.0          0.0   
21111          1   39          0.0          0.0          0.0          0.0   
...          ...  ...          ...          ...          ...          ...   
13123          1   90          0.0          0.0          0.0          0.0   
19648          1   36          0.0          1.0          0.0          0.0   
9845           1   26          0.0          1.0          0.0          0.0   
10799          1   44          0.0          1.0          0.0          0.0   
2732           1   39          0.0          1.0          0.0          0.0   

       workclass_4  workclass_5  workclass_6  workclass_7  ...  \
9281           0.0          0.0          0.0          0.0  ...   
31884          0.0          0.0          0.0          0.0  ...   
31580          0.0          0.0          0.0          0.0  ...   
18489          0.0          0.0          0.0          0.0  ...   
21111          0.0          1.0          0.0          0.0  ...   
...            ...          ...          ...          ...  ...   
13123          1.0          0.0          0.0          0.0  ...   
19648          0.0          0.0          0.0          0.0  ...   
9845           0.0          0.0          0.0          0.0  ...   
10799          0.0          0.0          0.0          0.0  ...   
2732           0.0          0.0          0.0          0.0  ...   

       native.country_31  native.country_32  native.country_33  \
9281                 0.0                0.0                0.0   
31884                0.0                0.0                0.0   
31580                0.0                0.0                0.0   
18489                0.0                0.0                0.0   
21111                0.0                0.0                0.0   
...                  ...                ...                ...   
13123                0.0                0.0                0.0   
19648                0.0                0.0                0.0   
9845                 0.0                0.0                0.0   
10799                0.0                0.0                0.0   
2732                 0.0                0.0                0.0   

       native.country_34  native.country_35  native.country_36  \
9281                 0.0                0.0                0.0   
31884                0.0                0.0                0.0   
31580                0.0                0.0                0.0   
18489                0.0                0.0                0.0   
21111                0.0                0.0                0.0   
...                  ...                ...                ...   
13123                0.0                0.0                0.0   
19648                0.0                0.0                0.0   
9845                 0.0                0.0                0.0   
10799                0.0                0.0                0.0   
2732                 0.0                0.0                0.0   

       native.country_37  native.country_38  native.country_39  \
9281                 0.0                0.0                0.0   
31884                0.0                0.0                0.0   
31580                0.0                0.0                0.0   
18489                0.0                0.0                0.0   
21111                0.0                0.0                0.0   
...                  ...                ...                ...   
13123                0.0                0.0                0.0   
19648                0.0                0.0                0.0   
9845                 0.0                0.0                0.0   
10799                0.0                0.0

In [ ]:
model = build_model(x_train.shape[1:])
model.fit(x=x_train, y=y_train, batch_size=batch_size, epochs=epochs, shuffle=shuffle, validation_data=(x_test, y_test))

Epoch 1/20
916/916 [==============================] - 10s 9ms/step - loss: 1.6068 - accuracy: 0.7683 - val_loss: 0.8291 - val_accuracy: 0.6620
Epoch 2/20
916/916 [==============================] - 7s 8ms/step - loss: 0.6221 - accuracy: 0.7818 - val_loss: 0.6185 - val_accuracy: 0.7651
Epoch 3/20
916/916 [==============================] - 7s 8ms/step - loss: 0.6026 - accuracy: 0.7852 - val_loss: 0.5778 - val_accuracy: 0.7875
Epoch 4/20
916/916 [==============================] - 7s 8ms/step - loss: 0.5871 - accuracy: 0.7896 - val_loss: 0.5873 - val_accuracy: 0.7734
Epoch 5/20
916/916 [==============================] - 7s 8ms/step - loss: 0.5833 - accuracy: 0.7884 - val_loss: 0.5877 - val_accuracy: 0.7872
Epoch 6/20
916/916 [==============================] - 7s 8ms/step - loss: 0.5797 - accuracy: 0.7871 - val_loss: 0.5746 - val_accuracy: 0.7746
Epoch 7/20
916/916 [==============================] - 7s 8ms/step - loss: 0.5740 - accuracy: 0.7880 - val_loss: 0.5663 - val_accuracy: 0.7866
Epoch

Best Validation Accuracy: **0.7974**, which is similar to the accuracy of dummy encoding technique.

# Binary Encoding of Inputs

Binary encoding is a combination of hash encoding and one hot encoding.

In [ ]:
from category_encoders.binary import BinaryEncoder

x_train, x_test, y_train, y_test = train_test_split(
    data,
    targets,
    test_size=test_size,
    random_state=random_state
)

binary_encoder = BinaryEncoder()
binary_encoder.fit(data)
x_train = binary_encoder.transform(x_train)
x_test = binary_encoder.transform(x_test)
x_train

age  workclass_0  workclass_1  workclass_2  workclass_3  fnlwgt  \
9281    35            0            0            1            0  241126   
31884   28            0            0            1            0   90547   
31580   55            0            0            1            0   70088   
18489   33            0            0            1            0  182423   
21111   39            0            1            1            0  163057   
...    ...          ...          ...          ...          ...     ...   
13123   90            0            1            0            1  282095   
19648   36            0            0            1            0  279721   
9845    26            0            0            1            0   51961   
10799   44            0            0            1            0  115323   
2732    39            0            0            1            0  224531   

       education_0  education_1  education_2  education_3  ...  sex_1  \
9281             0            1            0            0  ...      0   
31884            0            0            0            0  ...      1   
31580            0            1            0            0  ...      0   
18489            0            0            0            0  ...      0   
21111            0            0            0            1  ...      0   
...            ...          ...          ...          ...  ...    ...   
13123            0            0            0            1  ...      0   
19648            0            0            0            0  ...      0   
9845             0            1            1            1  ...      0   
10799            0            1            0            0  ...      0   
2732             0            0            0            0  ...      0   

       capital.gain  capital.loss  hours.per.week  native.country_0  \
9281              0             0              40                 0   
31884             0             0              23                 0   
31580             0             0              40                 0   
18489             0             0              40                 0   
21111             0             0              99                 0   
...             ...           ...             ...               ...   
13123             0             0              40                 0   
19648             0             0              40                 0   
9845              0             0              51                 0   
10799             0             0              40                 0   
2732           7298             0              40                 0   

       native.country_1  native.country_2  native.country_3  native.country_4  \
9281                  0                 0                 0                 0   
31884                 0                 0                 0                 0   
31580                 0                 0                 0                 0   
18489                 0                 0                 0                 0   
21111                 0                 0                 0                 0   
...                 ...               ...               ...               ...   
13123                 0                 0                 0                 0   
19648                 0                 0                 0                 0   
9845                  0                 0                 0                 0   
10799                 0                 0                 0                 0   
2732                  0                 0                 0                 0   

       native.country_5  
9281                  1  
31884                 1  
31580                 1  
18489                 1  
21111                 1  
...                 ...  
13123                 1  
19648                 1  
9845                  1  
10799                 1  
2732                  1  

[29304 rows x 36 columns]

In [ ]:
model = build_model(x_train.shape[1:])
model.fit(x=x_train, y=y_train, batch_size=batch_size, epochs=epochs, shuffle=shuffle, validation_data=(x_test, y_test))

Epoch 1/20
916/916 [==============================] - 11s 8ms/step - loss: 1.6675 - accuracy: 0.7724 - val_loss: 0.6335 - val_accuracy: 0.7906
Epoch 2/20
916/916 [==============================] - 7s 8ms/step - loss: 0.6336 - accuracy: 0.7812 - val_loss: 0.6054 - val_accuracy: 0.7854
Epoch 3/20
916/916 [==============================] - 7s 8ms/step - loss: 0.6015 - accuracy: 0.7846 - val_loss: 0.6418 - val_accuracy: 0.7307
Epoch 4/20
916/916 [==============================] - 7s 8ms/step - loss: 0.5899 - accuracy: 0.7870 - val_loss: 0.5834 - val_accuracy: 0.7906
Epoch 5/20
916/916 [==============================] - 7s 8ms/step - loss: 0.5835 - accuracy: 0.7879 - val_loss: 0.6220 - val_accuracy: 0.7467
Epoch 6/20
916/916 [==============================] - 7s 8ms/step - loss: 0.5784 - accuracy: 0.7860 - val_loss: 0.5688 - val_accuracy: 0.7869
Epoch 7/20
916/916 [==============================] - 8s 8ms/step - loss: 0.5715 - accuracy: 0.7898 - val_loss: 0.5726 - val_accuracy: 0.7866
Epoch

Best Validation Accuracy: **0.7906**

#Hash Encoder

Hashing involves transforming any given categorical variable into a fixed and unique numerical value using a hash function. Popular hash functions for Hash encoders include *MD, MD2, MD5*. In this example, we will use the **MD5** method. The advantage of the hash encoder over previous approaches is that it can use any number of additional columns to describe categories. 20 Additional featutures will be used to hash the categories of this dataset.

In [ ]:
from category_encoders.hashing import HashingEncoder

x_train, x_test, y_train, y_test = train_test_split(
    data,
    targets,
    test_size=test_size,
    random_state=random_state
)

hash_encoder = HashingEncoder(n_components=20, hash_method='md5')
hash_encoder.fit(data)
x_train = hash_encoder.transform(x_train)
x_test = hash_encoder.transform(x_test)
x_train

col_0  col_1  col_2  col_3  col_4  col_5  col_6  col_7  col_8  col_9  \
9281       0      0      0      0      0      0      0      0      1      0   
31884      0      0      0      1      1      0      0      0      0      0   
31580      0      0      0      0      0      0      0      0      1      0   
18489      0      0      0      0      1      1      0      0      0      0   
21111      1      0      0      0      0      1      1      0      0      0   
...      ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
13123      0      0      0      0      0      0      0      0      1      0   
19648      0      0      0      0      0      0      0      0      1      0   
9845       0      1      1      0      0      1      0      0      0      0   
10799      0      0      1      0      0      0      0      0      1      0   
2732       0      0      0      0      0      0      1      0      1      0   

       ...  col_16  col_17  col_18  col_19  age  fnlwgt  education.num  \
9281   ...       0       2       1       1   35  241126             14   
31884  ...       0       1       3       1   28   90547              9   
31580  ...       0       2       1       1   55   70088             14   
18489  ...       0       2       4       0   33  182423              9   
21111  ...       0       1       1       1   39  163057             10   
...    ...     ...     ...     ...     ...  ...     ...            ...   
13123  ...       0       2       1       2   90  282095             10   
19648  ...       0       2       2       2   36  279721              9   
9845   ...       0       2       2       0   26   51961              8   
10799  ...       0       2       1       1   44  115323             14   
2732   ...       0       2       2       1   39  224531              9   

       capital.gain  capital.loss  hours.per.week  
9281              0             0              40  
31884             0             0              23  
31580             0             0              40  
18489             0             0              40  
21111             0             0              99  
...             ...           ...             ...  
13123             0             0              40  
19648             0             0              40  
9845              0             0              51  
10799             0             0              40  
2732           7298             0              40  

[29304 rows x 26 columns]

In [ ]:
model = build_model(x_train.shape[1:])
model.fit(x=x_train, y=y_train, batch_size=batch_size, epochs=epochs, shuffle=shuffle, validation_data=(x_test, y_test))

Epoch 1/20
916/916 [==============================] - 9s 8ms/step - loss: 1.6417 - accuracy: 0.7719 - val_loss: 0.6505 - val_accuracy: 0.7928
Epoch 2/20
916/916 [==============================] - 7s 8ms/step - loss: 0.6241 - accuracy: 0.7830 - val_loss: 0.6141 - val_accuracy: 0.7869
Epoch 3/20
916/916 [==============================] - 7s 8ms/step - loss: 0.5997 - accuracy: 0.7880 - val_loss: 0.6015 - val_accuracy: 0.7839
Epoch 4/20
916/916 [==============================] - 7s 8ms/step - loss: 0.5897 - accuracy: 0.7868 - val_loss: 0.5934 - val_accuracy: 0.7875
Epoch 5/20
916/916 [==============================] - 7s 8ms/step - loss: 0.5805 - accuracy: 0.7892 - val_loss: 0.5760 - val_accuracy: 0.7832
Epoch 6/20
916/916 [==============================] - 7s 8ms/step - loss: 0.5733 - accuracy: 0.7892 - val_loss: 0.5601 - val_accuracy: 0.8032
Epoch 7/20
916/916 [==============================] - 7s 8ms/step - loss: 0.5751 - accuracy: 0.7887 - val_loss: 0.5671 - val_accuracy: 0.7881
Epoch 

Best Validation Accuracy: **0.8032**

#BaseN Encoder

Encoding the categorial variables using numbers from an arithmetic system of base "4".

In [ ]:
from category_encoders.basen import BaseNEncoder

x_train, x_test, y_train, y_test = train_test_split(
    data,
    targets,
    test_size=test_size,
    random_state=random_state
)

basen_encoder = BaseNEncoder(base=4)
basen_encoder.fit(data)
x_train = basen_encoder.transform(x_train)
x_test = basen_encoder.transform(x_test)
x_train

age  workclass_0  workclass_1  fnlwgt  education_0  education_1  \
9281    35            0            2  241126            0            2   
31884   28            0            2   90547            0            0   
31580   55            0            2   70088            0            2   
18489   33            0            2  182423            0            0   
21111   39            1            2  163057            0            0   
...    ...          ...          ...     ...          ...          ...   
13123   90            1            1  282095            0            0   
19648   36            0            2  279721            0            0   
9845    26            0            2   51961            0            3   
10799   44            0            2  115323            0            2   
2732    39            0            2  224531            0            0   

       education_2  education.num  marital.status_0  marital.status_1  ...  \
9281             0             14                 1                 1  ...   
31884            1              9                 1                 1  ...   
31580            0             14                 1                 1  ...   
18489            1              9                 0                 2  ...   
21111            2             10                 0                 2  ...   
...            ...            ...               ...               ...  ...   
13123            2             10                 1                 1  ...   
19648            1              9                 1                 1  ...   
9845             2              8                 1                 0  ...   
10799            0             14                 1                 1  ...   
2732             1              9                 1                 1  ...   

       relationship_1  race_0  race_1  sex_0  capital.gain  capital.loss  \
9281                1       0       1      2             0             0   
31884               2       0       2      1             0             0   
31580               1       0       1      2             0             0   
18489               2       0       2      2             0             0   
21111               1       0       1      2             0             0   
...               ...     ...     ...    ...           ...           ...   
13123               1       0       1      2             0             0   
19648               1       0       1      2             0             0   
9845                0       0       2      2             0             0   
10799               1       0       1      2             0             0   
2732                1       0       1      2          7298             0   

       hours.per.week  native.country_0  native.country_1  native.country_2  
9281               40                 0                 0                 1  
31884              23                 0                 0                 1  
31580              40                 0                 0                 1  
18489              40                 0                 0                 1  
21111              99                 0                 0                 1  
...               ...               ...               ...               ...  
13123              40                 0                 0                 1  
19648              40                 0                 0                 1  
9845               51                 0                 0                 1  
10799              40                 0                 0                 1  
2732               40                 0                 0                 1  

[29304 rows x 23 columns]

In [ ]:
model = build_model(x_train.shape[1:])
model.fit(x=x_train, y=y_train, batch_size=batch_size, epochs=epochs, shuffle=shuffle, validation_data=(x_test, y_test))

Epoch 1/20
916/916 [==============================] - 11s 8ms/step - loss: 1.5810 - accuracy: 0.7698 - val_loss: 0.8075 - val_accuracy: 0.6945
Epoch 2/20
916/916 [==============================] - 7s 8ms/step - loss: 0.6178 - accuracy: 0.7839 - val_loss: 0.7165 - val_accuracy: 0.6524
Epoch 3/20
916/916 [==============================] - 8s 8ms/step - loss: 0.6009 - accuracy: 0.7823 - val_loss: 0.5845 - val_accuracy: 0.7710
Epoch 4/20
916/916 [==============================] - 9s 10ms/step - loss: 0.5899 - accuracy: 0.7833 - val_loss: 0.6068 - val_accuracy: 0.7860
Epoch 5/20
916/916 [==============================] - 7s 8ms/step - loss: 0.5821 - accuracy: 0.7823 - val_loss: 0.6083 - val_accuracy: 0.7614
Epoch 6/20
916/916 [==============================] - 7s 8ms/step - loss: 0.5706 - accuracy: 0.7872 - val_loss: 0.5674 - val_accuracy: 0.7808
Epoch 7/20
916/916 [==============================] - 7s 8ms/step - loss: 0.5676 - accuracy: 0.7863 - val_loss: 0.5960 - val_accuracy: 0.7611
Epoc

Best Validation Accuracy: **0.7921**

#Other Categorical Encoders

In [17]:
from category_encoders.backward_difference import BackwardDifferenceEncoder
from category_encoders.cat_boost import CatBoostEncoder
from category_encoders.count import CountEncoder
from category_encoders.glmm import GLMMEncoder
from category_encoders.target_encoder import TargetEncoder
from category_encoders.helmert import HelmertEncoder
from category_encoders.james_stein import JamesSteinEncoder

encoders = {
    'Backward Difference Encoder': BackwardDifferenceEncoder(),
    'Cat Boost Encoder': CatBoostEncoder(),
    'Count Encoder': CountEncoder(),
    'Generalized Linear Mixed Model Encoder': GLMMEncoder(binomial_target=True),
    'Target Encoder': TargetEncoder(),
    'Helmert Encoder': HelmertEncoder(),
    'James-Stein Encoder': JamesSteinEncoder()
}

for encoder_name, encoder in encoders.items():
  print('Evaluating: {}'.format(encoder))
  x_train, x_test, y_train, y_test = train_test_split(
    data,
    targets,
    test_size=test_size,
    random_state=random_state
  )

  encoder.fit(data, targets)
  x_train = encoder.transform(x_train)
  x_test = encoder.transform(x_test)
  
  model = build_model(x_train.shape[1:])
  model.fit(x=x_train, y=y_train, batch_size=batch_size, epochs=epochs, shuffle=shuffle, validation_data=(x_test, y_test))

Evaluating: BackwardDifferenceEncoder()
Epoch 1/20
916/916 [==============================] - 7s 4ms/step - loss: 1.6122 - accuracy: 0.7715 - val_loss: 0.9418 - val_accuracy: 0.6742
Epoch 2/20
916/916 [==============================] - 3s 3ms/step - loss: 0.6321 - accuracy: 0.7838 - val_loss: 0.7613 - val_accuracy: 0.6380
Epoch 3/20
916/916 [==============================] - 3s 3ms/step - loss: 0.6051 - accuracy: 0.7862 - val_loss: 0.6074 - val_accuracy: 0.7811
Epoch 4/20
916/916 [==============================] - 3s 3ms/step - loss: 0.5926 - accuracy: 0.7853 - val_loss: 0.5854 - val_accuracy: 0.7835
Epoch 5/20
916/916 [==============================] - 3s 3ms/step - loss: 0.5846 - accuracy: 0.7857 - val_loss: 0.5915 - val_accuracy: 0.7777
Epoch 6/20
916/916 [==============================] - 3s 3ms/step - loss: 0.5760 - accuracy: 0.7887 - val_loss: 0.5742 - val_accuracy: 0.7814
Epoch 7/20
916/916 [==============================] - 3s 3ms/step - loss: 0.5713 - accuracy: 0.7882 - val_lo